### ISSUE TITLE
Mobile iOS Version will not load or save patches

### ISSUE THREAD

**Issue by o-g-sus** [plugdata-team/plugdata#1375]
I quickly tested the mobile Version of Plugdata, here is my feedback:
- Load & Save Patches does not work yet. Browser window opens, but not Patch will open, nor you can save a Patch on Mobile

**Comment by timothyschoen** [comment]
> * Load & Save Patches does not work yet. Browser window opens, but not Patch will open, nor you can save a Patch on Mobile

Kind of, it does work, but iOS doesn't allow you to open/close files from any location for security reasons. It will only allow you to open/close patches if they are in the "plugdata" folder in documents.

So right now, the best workflow is to airdrop your patches over, save them in the "plugdata" folder, and work from there. You should also be able to save inside the plugdata folder. I'm not sure if there is any workaround for this, I couldn't find any.


> * Additionally it would be great, if we could use the handy OSX/iOS way to share an import documents (PD patches) via Airdrop, etc. ... share ... open in: https://stackoverflow.com/questions/3981199/adding-open-in-option-to-ios-app

I agree, I'll work on that!

**Comment by o-g-sus** [comment]
Yeah, found out about the workaround already :)
 
 "It will only allow you to open/close patches if they are in the "plugdata" folder in documents." - That's correct.
 
 But with the "open in" function you can load PD files from other locations of the device then also

**Comment by timothyschoen** [comment]
I suppose this is kind of fixed? It was only that the label is named "move" right?

I agree that that's a weird name for the button, but I'm not sure how to solve it

**Comment by 60-hz** [comment]
I guess that "save to" might be more meaningful.

**Comment by o-g-sus** [comment]
You can find no "move" Sting in the Project? 
iOS branch is in develop right?

**Comment by Carnarts** [comment]
Hi, thanks for the update today :) Definitely working better in the UI on iOS AUv3 (although still some of the dropdowns in the settings can't be opened it seems)
But main issue is I'm having problems opening any pd patches which aren't self contained and need other files. Whatever I tried (also moving files further up the folder tree etc) doesn't work for me yet, I get a string of errors in the console. I don't think this was a problem before, the patches could be opened without those errors.

**Comment by timothyschoen** [comment]
> Hi, thanks for the update today :) Definitely working better in the UI on iOS AUv3 (although still some of the dropdowns in the settings can't be opened it seems) But main issue is I'm having problems opening any pd patches which aren't self contained and need other files. Whatever I tried (also moving files further up the folder tree etc) doesn't work for me yet, I get a string of errors in the console. I don't think this was a problem before, the patches could be opened without those errors.

Correct, I still need to fix that. I only fixed files access in plugdata's own source code, but I didn't fix it yet within Pd itself. So actions that are handled by plugdata (like opening/saving patches) work on iOS, but everything handled from within Pd (like abstractions, audio, images) does not.

I wonder how PdParty is able to access patch resources from Pd? Maybe @danomatika knows?

**Comment by danomatika** [comment]
libpd_add_searchpath()?enohp ym morf tnes-----------Dan Wilcoxdanomatika.comrobotcowboy.comOn Jan 31, 2024, at 8:37 PM, Timothy Schoen ***@***.***> wrote:﻿

Hi, thanks for the update today :) Definitely working better in the UI on iOS AUv3 (although still some of the dropdowns in the settings can't be opened it seems) But main issue is I'm having problems opening any pd patches which aren't self contained and need other files. Whatever I tried (also moving files further up the folder tree etc) doesn't work for me yet, I get a string of errors in the console. I don't think this was a problem before, the patches could be opened without those errors.

Correct, I still need to fix that. I only fixed files access in plugdata's own source code, but I didn't fix it yet within Pd itself. So actions that are handled by plugdata (like opening/saving patches) work on iOS, but everything handled from within Pd (like abstractions, audio, images) does not.
I wonder how PdParty is able to access patch resources from Pd? Maybe @danomatika knows?

**Comment by TricksterSam** [comment]
Was wondering if you’ve made any progress on this. I just got an iPad and was delighted I could use PlugData! Gotta be able to use my abstractions!

**Comment by danomatika** [comment]
> I wonder how PdParty is able to access patch resources from Pd? Maybe [@danomatika](https://github.com/danomatika) knows?

Nothing special, just add the dir to the search path. There is an internal set of abstractions and support for a general `Documents/lib` path so those are added as well (recursively): https://github.com/danomatika/PdParty/blob/a55bfe8374f509b4bf557e579e30afa2fce73736/src/scenes/PatchScene.m#L35

**Comment by danomatika** [comment]
> So actions that are handled by plugdata (like opening/saving patches) work on iOS, but everything handled from within Pd (like abstractions, audio, images) does not.

Messages are used for other actions between libpd and the GUI implementation in PdParty. For instance, I copied droidparty's [loadsave] object. This is a dummy abstraction using [openpanel] and [savepanel] when patching on desktop but in PdParty this is an abstraction which is sends a message form the libpd instance to the Obj-C layer which sets the dir and opens the a native panel also written in Obj-C.

**Comment by JulienVincenot** [comment]
Hi !

Sorry to revive an old topic

I'm an old time Max user and I just discovered plugdata a few days ago, project looks fantastic ! I will definitely send my students your way for intro classes :) I just happened today to try it on iPad which feels… surreal, especially with mouse and keyboard. Still a few rough edges here but it's just too welcoming for Max users to be a deal breaker :)

One thing that was not so obvious to me was that saving patch permission issue. I eventually found the solution here, but I'm sure beginners would get clueless and perhaps give up… Would it be possible to document that with some popup the first time plugdata tries to save on that platform?

My main question is the following : would it be imaginable on iOS to have similar permissions for a plugdata folder but inside the iCloud folder instead? I personally never use the local one ("On My iPad") and iCloud is really the one place I can rely on easily accross devices, Mac, phone etc.

Thanks !

**Comment by danomatika** [comment]
> I'm not sure to understand why you talk about PlugData as a plugin, it is an app no?

I don't know as I haven't used Plugdata.

From the discussion involving "AUv3" I assume it was bundling people's patches into an audio unit to be loaded in other apps.

If it is indeed an iOS app as well, then I probably can't add much to how it works other than to comment on how it can organize itself to be able to load file paths.

**Comment by JulienVincenot** [comment]
@danomatika I never talked about AUv3. Better wait for @timothyschoen , I'm sure he has used Plugdata  😅

**Comment by danomatika** [comment]
I'm referring to the original post from 2024 which included me here. In any case, sorry for the noise. I will leave the conversation then.

**Comment by JulienVincenot** [comment]
Sorry that's on me, I only read the first messages, I assumed there was only one way to get this to work on iOS 😭
